p2p/kademlia.py

```diff
@@ -220,11 +221,29 @@ def __lt__(self, other: 'KBucket') -> bool:
 
 
 class RoutingTable:
+    logger = logging.getLogger("p2p.kademlia.RoutingTable")
 
     def __init__(self, node: Node) -> None:
         self.this_node = node
         self.buckets = [KBucket(0, k_max_node_id)]
 
+    def get_random_nodes(self, count: int) -> Generator[Node, None, None]:
+        if count > len(self):
+            self.logger.warn(
+                "Cannot get %d nodes as RoutingTable contains only %d nodes", count, len(self))
+            count = len(self)
+        seen = []  # type: List[Node]
+        # This is a rather inneficient way of randomizing nodes from all buckets, but even if we
+        # iterate over all nodes in the routing table, the time it takes would still be
+        # insignificant compared to the time it takes for the network roundtrips when connecting
+        # to nodes.
+        while len(seen) < count:
+            bucket = random.choice(self.buckets)
+            node = random.choice(bucket.nodes)
```

Comment:
I did some experiments and *if* we make `RountingTable` inherit from `collections.abc.Sequence` **and** we add a `__getitem__` implementation (required to inherit from `Sequence`), then we can replace this whole loop with `random.sample(self, count)`

Reply:
Looking at the internals of `random.sample` we won't get any performance gains from this and are likely to take a performance hit (a little).  Easier to read though... I'm now leaning towards `-1` on using `random.sample` here.

However, I do think this implementation needs to be changed as this could take a **lot** of iterations if `len(all_bucket_nodes)` is large and `count` is also large as it will start to encounter more and more nodes it's already seen.

I came up with this solution.  Let me know what you think.

```python
def get_random_sample(self, count):
    all_node_idx_paths = [
        (bucket_idx, node_idx)
        for bucket_idx in range(len(self.buckets))
        for node_idx in range(len(self.buckets[bucket_idx].nodes))
    ]
    random.shuffle(all_node_idx_paths)
    selected_node_idx_paths = all_node_idx_paths[:count]
    return [
        self.buckets[bucket_idx].nodes[node_idx]
        for bucket_idx, node_idx
        in selected_node_idx_paths
    ]
```

Reply:
Well, this method is a generator and is used in a loop where, on each iteration it will attempt to connect and perform two handshakes with the remote (which means several network round trips that may take up to 10s with the current timeouts we're using). Also, we break out of the loop that uses it as soon as we've connected to `PeerPool.min_peers` peers, so we're very unlikely to consume so many items as to cause it to slow down. Given that, I'd rather keep this code as simple as possible for now, but I'd be ok with using your solution if you disagree

Reply:
Somehow I missed that it was a generator and thus that it would exit early.  👍 on leaving it as-is.